osu.Server.Queues.ScoreStatisticsProcessor/Processors/MedalAwarders/PackMedalAwarder.cs

```diff
@@ -267,17 +268,31 @@ bool checkPack(int packId, bool noReductionMods = false)
                     if (context.Score.RulesetID != packRulesetId)
                         return false;
 
-                    rulesetCriteria = $"AND ruleset_id = {packRulesetId}";
+                    selectClauses += $" AND s.ruleset_id = {packRulesetId}";
                 }
 
-                // TODO: no index on (beatmap_id, user_id) may mean this is too slow.
-                // note that the `preserve = 1` condition relies on the flag being set before score processing (https://github.com/ppy/osu-web/pull/10946).
-                int completed = context.Connection.QuerySingle<int>(
-                    "SELECT COUNT(distinct p.beatmapset_id)"
-                    + "FROM osu_beatmappacks_items p "
-                    + "JOIN osu_beatmaps b USING (beatmapset_id) "
-                    + "JOIN scores s USING (beatmap_id)"
-                    + $"WHERE s.user_id = {context.Score.UserID} AND s.passed = 1 AND s.preserve = 1 AND pack_id = {packId} {rulesetCriteria} {modsCriteria}", transaction: context.Transaction);
+                int completed;
+
+                if (noReductionMods)
+                {
+                    var scoresWithBeatmapsetId = context.Connection.Query(
+                        $"SELECT s.data, p.beatmapset_id {selectClauses}",
+                        (SoloScore score, uint beatmapsetId) => (score, beatmapsetId),
+                        transaction: context.Transaction,
+                        splitOn: "beatmapset_id");

```

Comment:
I don't have production access to check empirically but I am expecting that pulling all user scores like that is going to be an issue due to the volume of data involved.

Reply:
This is also gonna be a problem [here](https://github.com/ppy/osu-web/blob/b29eeff28e100bd5ed43a26f355575f19918e556/app/Models/BeatmapPack.php#L118-L120).

Reply:
I'd feel much more comfortable with the pp based solution for now. As a separate PR which only does that. This can be left in draft for future consideration.

Goal would be to fix this immediately and dwell on what this PR is attempting to do. I do not feel comfortable even beginning to investigate the performance concerns of this PR at this stage.

Reply:
can't you just set the pp manually in the test?

Reply:
https://github.com/ppy/osu-queue-score-statistics/pull/234

Reply:
One corollary of that temporary change is that loved packs won't work for the time being (because loved scores don't get pp). It is what it is for now I guess...

Reply:
thx :pray: 

for loved, there are currently no challenge packs, I'll keep this in mind if it ever comes up